### DISCUSSION TITLE
Configure `charon-cmd` to use ECDSA key

### DISCUSSION THREAD

**Discussion by kwohlfahrt** [strongswan/strongswan#1955] [Q&A]
I have an ECDSA key I want to use with `charon-cmd`. How to I configure `charon-cmd` to load this?

So far, I'm only familiar with the `swanctl.conf` style of configuration, but this doesn't seem to be read by `charon-cmd`. The command-line flags only include `--rsa` for the private key (and passing my ECDSA key here doesn't work). I also haven't been able to produce a PKCS12 file that works without a password for use with the `--p12` options.

For reference, the `swanctl.conf` that works is (apart from the domain names):

```
connections {
  ci {
    version = 2
    remote_addrs = vpn.example.org
    vips = ::,0.0.0.0

    local {
      auth = pubkey
      id = ci@example.org
    }
  
    remote {
      auth = pubkey
      id = vpn.example.org
    }

    children {
      ci {
        remote_ts = 192.168.32.0/24
      }
    }
  }
}
```

When running `charon-cmd --remote-ts 192.168.32.0/24 --host vpn.example.org --identity ci@example.org --cert /etc/strongswan/swanctl/x509ca/ca.pem`, I get an error that no certificates were found, followed by `AUTHENTICATION_FAILED`. I'm not clear how to tell `charon-cmd` where to find my ECDSA private key.

**Comment by tobiasbrunner** [comment]
Why use `charon-cmd` if you have a working swanctl.conf? And no, the latter is not read by `charon-cmd`, which only reads the global strongswan.conf and gets all connection-specific settings via command line arguments.

Anyway, `charon-cmd` does currently only support loading RSA keys unless the private key is contained in a PKCS#12 container. And yes, password-less PKCS#12 containers currently don't work (I pushed some patches for that to the _1955-pkcs12-no-pw_ branch and also added a commit that adds support for arbitrary private keys to `charon-cmd`).

**Reply by kwohlfahrt** [reply]
> Why use charon-cmd if you have a working swanctl.conf? And no, the latter is not read by charon-cmd, which only reads the global strongswan.conf and gets all connection-specific settings via command line arguments.

This is to create a VPN from a temporary environment (a CI runner specifically), which doesn't have systemd or another service manager. So running and backgrounding `charon-cmd` seemed like a sensible thing to do. I could also presumably start and background the `charon` daemon, and tell it to set up the connection with `swanctl`, but `charon-cmd` seemed like a more natural fit for a command to set up a connection.

I only set up the `swanctl.conf` after `charon-cmd` was clearly not working, just to sanity check my set-up.
